pymeasure/instruments/commonBase.py

```diff
@@ -0,0 +1,468 @@
+    """
+
+    # Variable holding the list of DynamicProperty parameters that are configurable
+    # by users
+    _fget_params_list = ('get_command',
+                         'values',
+                         'map_values',
+                         'get_process',
+                         'command_process',
+                         'check_get_errors')
+
+    _fset_params_list = ('set_command',
+                         'validator',
+                         'values',
+                         'map_values',
+                         'set_process',
+                         'command_process',
+                         'check_set_errors')
+
+    # Prefix used to store reserved variables
+    __reserved_prefix = "___"
+
+    def __init__(self):
+        self._special_names = self._setup_special_names()
+
+    def _setup_special_names(self):
+        """ Return list of class/instance special names.
+
+        Compute the list of special names based on the list of
+        class attributes that are a DynamicProperty. Check also for class variables
+        with special name and copy them at instance level
+        Internal method, not intended to be accessed at user level."""
+        special_names = []
+        dynamic_params = tuple(set(self._fget_params_list + self._fset_params_list))
+        # Check whether class variables of DynamicProperty type are present
+        for obj in (self,) + self.__class__.__mro__:
+            for attr_name, attr in obj.__dict__.items():
+                if isinstance(attr, DynamicProperty):
+                    special_names += [attr_name + "_" + key for key in dynamic_params]
+        # Check if special variables are defined at class level
+        for obj in (self,) + self.__class__.__mro__:
+            for attr in obj.__dict__:
+                if attr in special_names:
+                    # Copy class special variable at instance level, prefixing reserved_prefix
+                    setattr(self, self.__reserved_prefix + attr, obj.__dict__[attr])
+        return special_names
+
+    def __setattr__(self, name, value):
+        """ Add reserved_prefix in front of special variables."""
+        if hasattr(self, '_special_names'):
+            if name in self._special_names:
+                name = self.__reserved_prefix + name
+        super().__setattr__(name, value)
+
+    def __getattribute__(self, name):
+        """ Prevent read access to variables with special names used to
+        support dynamic property behaviour."""
+        if name in ('_special_names', '__dict__'):
+            return super().__getattribute__(name)
+        if hasattr(self, '_special_names'):
+            if name in self._special_names:
+                raise AttributeError(
+                    f"{name} is a reserved variable name and it cannot be read")
+        return super().__getattribute__(name)
+
+    # Channel management
+    def add_child(self, cls, id, group="channels", prefix="ch", **kwargs):
```

Comment:
I would not make the "channels" attribute name user-configurable, to be honest. The prefix will enable the customisation around `relay_1` etc, and it would be good to be able to rely on the fact that the bag of children (typically channels, but also channel-like "functions") is always called `instr.channels`. Or what use case are you trying to cover, here?

Reply:
> What use case are you trying to cover, here? Enabling a `instr.channels` + `instr.relays` for separate looping?

Exactly. Or analog channels and digital channels etc.

Looking from the perspective of the list as a complete list (forming the tree), your suggestion makes sense. What tree walking work do you expect or imagine, which is not implemented at instrument level (which knows the configuration)?
From the user perspective, you want to loop through a single group without remembering how many entries are from one type or the other. If we do not do it, the users will add lists containing their channels.

We could add a private list `_children` containing all children

Or we create a list channels, as you suggest, which contains all children. If an instrument has different types (not that often, I think), the implementor might create those separate lists himself. In the default case of several, identical channels, the default implementation remains sufficient.

This last approach is consistent with our current philosophy (as far as I read the most recent discussions): Cover basic usage in the pymeasure model (`Instrument`) and leave the edge cases to implementors. This keeps the code simple (not eierlegende Wollmilchsau), yet powerful.

Reply:
> What tree walking work do you expect or imagine, which is not implemented at instrument level (which knows the configuration)?

I don't know yet, but don't want to shut that door. However...

> We could add a private list _children containing all children

...this is something we can easily do later, should the need arise, **if** we require/document that children are added only through the `add_child` class. Then, we can handle any additional bookeeping in that class. 
For this to work, your interface as implemented now works fine, afaict -- the default case (bag of channels) only needs parentclass and `id`, anything else we currently know of is easily achievable.

To maintain API homogeneity across all instruments, we _should_ document that users should not deviate from the default prefix without good reasons.

What do you think about calling `group` `collection`, instead? `group` felt a little unclear to me, collection is a bit better, I'm open for even better proposals.

Reply:
I like collection.

Whit this, I'll write some documentation, which will propose the collection and prefix for special cases only (different types of channels).

I plan to write this information in the `add_child` docstring (regarding the parameters) as well.